contrib/pkg/labelobjects/labelobjects_test.go

```diff
@@ -0,0 +1,1188 @@
+	"github.com/openshift/hive/pkg/test/generic"
+	"github.com/openshift/hive/pkg/test/job"
+	"github.com/openshift/hive/pkg/test/persistentvolumeclaim"
+	"github.com/openshift/hive/pkg/test/secret"
+
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+const (
+	namespace = "notarealns"
+)
+
+var (
+	emptyRuntimeObjectSlice = []runtime.Object{}
+)
+
+func fakeLabelMigrator(existingObjects []runtime.Object) *labelMigrator {
+	return &labelMigrator{
+		client: fake.NewFakeClient(existingObjects...),
+		logger: log.WithField("type", "fake"),
+	}
+}
+
+func fakeClusterDeployment() *hivev1alpha1.ClusterDeployment {
+	return &hivev1alpha1.ClusterDeployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "fakeclusterdeployment",
+			Namespace: namespace,
+		},
+	}
+}
+
+func fakeClusterDeprovisionRequest() *hivev1alpha1.ClusterDeprovisionRequest {
+	return &hivev1alpha1.ClusterDeprovisionRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "fakeclusterdeprovisionrequest",
+			Namespace: namespace,
+		},
+	}
+}
+
+func fakeClusterProvision() *hivev1alpha1.ClusterProvision {
+	return &hivev1alpha1.ClusterProvision{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "fakeclusterprovision",
+			Namespace: namespace,
+		},
+	}
+}
+
+func fakeSyncSet() *hivev1alpha1.SyncSet {
+	return &hivev1alpha1.SyncSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "fakesyncset",
+			Namespace: namespace,
+		},
+	}
+}
+
+func fakeSelectorSyncSet() *hivev1alpha1.SelectorSyncSet {
+	return &hivev1alpha1.SelectorSyncSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "fakeselectorsyncset",
+			Namespace: namespace,
+		},
+	}
+}
+
+func TestRun(t *testing.T) {
+	apis.AddToScheme(scheme.Scheme)
+
+	tests := []struct {
+		name                 string
+		existingObjects      []runtime.Object
+		expectedError        error
+		expectedSuccessCount int
+		expectedFailureCount int
+	}{
+		{
+			name:            "Empty Namespace",
+			existingObjects: emptyRuntimeObjectSlice,
+		},
+		{
+			name: "1 ClusterProvision",
+			existingObjects: []runtime.Object{
+				buildClusterProvision(clusterProvisionBase(), genericClusterProvision(generic.WithControllerOwnerReference(fakeClusterDeployment()))),
+				persistentvolumeclaim.Build(persistentVolumeClaimBase(),
+					persistentvolumeclaim.Generic(generic.WithLabel(constants.InstallJobLabel, "true")),
+					persistentvolumeclaim.Generic(generic.WithControllerOwnerReference(fakeClusterDeployment()))),
+				job.Build(
+					jobBase(),
+					job.Generic(generic.WithNamePostfix("1")),
+					job.Generic(generic.WithLabel(hiveimageset.ImagesetJobLabel, "true")),
+					job.Generic(generic.WithControllerOwnerReference(fakeClusterDeployment()))),
```

Comment:
To follow this I cloned the branch and clicked through roughly 6 functions to see what was going on to build this test job.

For discussion purposes, compare to:

```go
testJob(
    "somejob-1", 
    namespace, 
    map[string]string{hiveimageset.ImagesetJobLabel: "true"}, 
    fakeClusterDeployment())
```

Maybe it's just because this is how we typically create these test objects but I find this clearer and easier to understand without jumping around very far. I feel like these builder functions are just another way to do things with roughly the same amount of code, more indirection, and another process the reader has to understand to work with it.

Granted it does mean "testJob" functions littered throughout various test modules.

What do we think, does this provide benefits I'm missing?

Reply:
I see what you're saying, but I think that may just be a familiarity thing. It took me awhile to feel comfortable with it when @staebler suggested using it, but now it's second nature.

The problem with what you listed is that the `testJob` function parameters aren't flexible (you have a fixed number of params that must be passed in).

For example, if you wanted some jobs to have a label and some other jobs to have a controller reference, but neither to have both, you'd need multiple `testJob` functions.

With the builder pattern on the other hand, you just put in whatever builder functions you want and the object is created exactly as you've asked.

The builders are also very easy to extend without affecting existing users. Alternatively if you change the signature of the `testJob` function, you must now go change all callers and possibly declare more variables to pass in to the function.

When building the extensions, they can also live out in a package that all tests can import or alternatively if it's something just for your tests in this one package, you can just include the builder function in that one package.

It's really nice to only have to code the builder functions once, and then we can use them in any of our tests to build up any type of object that we need.

Also, using the `job.Generic` pattern allows me to use a bunch of general functions that work on any object that implements `metav1.Object` (which all core and hive objects do).

When we originally talked about using a pattern like this, we discussed using either this syntax (which we got from the pattern used for list options in controller runtime), or possibly using a fluent pattern (like what logrus uses).

Here's roughly the equivalent in a fluent syntax:
```golang
jobBuilder
    .Base()
    .WithNamePostfix("1")
    .WithLabel(hiveimageset.ImagesetJobLabel, "true")
    .WithControllerOwnerReference(fakeClusterDeployment())
```

Would this have made more sense?

In other words, is the objection to having the builder pattern, or is it the syntax of the builders?

Reply:
The objection was to using a builder pattern in general, which I generally consider for construction of complex objects, and I don't see the go structs in our tests as qualifying for that when were setting metadata and labels. Personally I would stick to the existing methods, or inline the structs if we need a lot of flexibility. 

This is good context above though and if you both are on board, based it off established controller runtime patterns and enjoy using it then I'm ok with it.